ygen/directory.go

```diff
@@ -133,6 +137,124 @@ func GetOrderedDirectories(directory map[string]*Directory) ([]string, map[strin
+// Directory names and a map of Directories keyed by their paths, so that each
+// directory can be processed in path-alphabetical order. This helps produce
+// deterministic generated code, and minimize diffs when compared with expected
+// output (i.e., diffs don't appear simply due to reordering of the Directory
+// maps).
+func GetOrderedPathDirectories(directory map[string]*Directory) []string {
+	orderedDirPaths := make([]string, 0, len(directory))
+
+	for path := range directory {
+		orderedDirPaths = append(orderedDirPaths, path)
+	}
+	sort.Strings(orderedDirPaths)
+
+	return orderedDirPaths
+}
+
+// getOrderedDirDetails takes in a language-specific LangMapper, a map of
+// Directory objects containing the raw AST information, a schemaTree, and IR
+// generation options, and returns a map of ParsedDirectory objects that form
+// the primary component of ygen's IR output.
+func getOrderedDirDetails(langMapper LangMapper, directory map[string]*Directory, schematree *schemaTree, opts IROptions) (map[string]*ParsedDirectory, error) {
+	dirDets := map[string]*ParsedDirectory{}
+	for _, dirPath := range GetOrderedPathDirectories(directory) {
+		dir := directory[dirPath]
+		packageName, err := langMapper.PackageName(dir.Entry, opts.TransformationOptions.CompressBehaviour, opts.NestedDirectories)
+		if err != nil {
+			return nil, err
+		}
+		pd := &ParsedDirectory{
+			Name:        dir.Name,
+			PackageName: packageName,
+			IsFakeRoot:  dir.IsFakeRoot,
+			entry:       dir.Entry,
+		}
+		switch {
+		case dir.Entry.IsList():
+			pd.Type = List
+			pd.ListKeys = dir.ListAttr.Keys
+		default:
+			pd.Type = Container
+		}
+
+		pd.Fields = make(map[string]*NodeDetails, len(dir.Fields))
+		for _, fn := range GetOrderedFieldNames(dir) {
+			field := dir.Fields[fn]
+
+			mp, mm, err := findMapPaths(dir, fn, opts.TransformationOptions.CompressBehaviour.CompressEnabled(), false, opts.AbsoluteMapPaths)
+			if err != nil {
+				return nil, err
+			}
+
+			smp, smm, err := findMapPaths(dir, fn, opts.TransformationOptions.CompressBehaviour.CompressEnabled(), true, opts.AbsoluteMapPaths)
+			if err != nil {
+				return nil, err
+			}
+
+			mod, err := field.InstantiatingModule()
+			if err != nil {
+				return nil, err
+			}
+
+			var target *yang.Entry
+			if field.Type != nil && field.Type.Kind == yang.Yleafref {
+				if target, err = schematree.resolveLeafrefTarget(field.Type.Path, field); err != nil {
+					return nil, fmt.Errorf("unable to resolve leafref field: %v", err)
+				}
+			}
+
+			name, err := langMapper.FieldName(field)
+			if err != nil {
+				return nil, err
+			}
+
+			nd := &NodeDetails{
+				Name: name,
+				YANGDetails: YANGNodeDetails{
+					Name:         field.Name,
+					Defaults:     field.DefaultValues(),
+					Module:       mod,
+					Path:         field.Path(),
+					ResolvedPath: target.Path(),
+				},
+				MappedPaths:         mp,
+				MappedModules:       mm,
```

Comment:
Question here, why `MappedModules`? I find the name here it a little confusing, since it's not clear that we "map" to a module - just that a node is defined by that module, right?

The "map" in paths was because this field is "mapped" to one or more output YANG paths (just remembering what the original intent there was).

Reply:
I'll change it to `MappedPathModules`. I named it `MappedModules` because of congruency, so that it's clear to the user these two are related, which I think might be the quickest way for the reader to understand their meanings.

I was considering what the naming should be here. I actually wanted to merge the two strings into a single struct but it was too cumbersome to change all the existing code, and I found I frequently needed to convert to `[]string` for processing. I could define a type to do that but I thought keeping it like this was simpler so I left it be.

Reply:
SGTM.